packages/datadog-shimmer/src/shimmer.js

```diff
@@ -0,0 +1,137 @@
+'use strict'
+
+// Use a weak map to avoid polluting the wrapped function/method.
+const unwrappers = new WeakMap()
+
+function copyProperties (original, wrapped) {
+  Object.setPrototypeOf(wrapped, original)
+
+  const props = Object.getOwnPropertyDescriptors(original)
+  const keys = Reflect.ownKeys(props)
+
+  for (const key of keys) {
+    Object.defineProperty(wrapped, key, props[key])
+  }
+}
+
+function wrapFn (original, delegate) {
+  assertFunction(delegate)
+  assertNotClass(original) // TODO: support constructors of native classes
+
+  const shim = function shim () {
+    return delegate.apply(this, arguments)
+  }
+
+  unwrappers.set(shim, () => {
+    delegate = original
+  })
+
+  copyProperties(original, shim)
+
+  return shim
+}
+
+function wrapMethod (target, name, wrapper) {
+  assertMethod(target, name)
+  assertNotClass(target[name]) // TODO: support constructors of native classes
+  assertFunction(wrapper)
+
+  const original = target[name]
+  const wrapped = wrapper(original)
+  const descriptor = Object.getOwnPropertyDescriptor(target, name)
+
+  if (descriptor) {
+    unwrappers.set(wrapped, () => Object.defineProperty(target, name, descriptor))
+  } else { // no descriptor means original was on the prototype
+    unwrappers.set(wrapped, () => delete target[name])
+  }
+
+  Object.defineProperty(target, name, {
+    configurable: true,
+    writable: true,
+    enumerable: false,
+    ...descriptor,
+    value: wrapped
+  })
+
+  copyProperties(original, wrapped)
+
+  return target
+}
+
+function wrap (target, name, wrapper) {
+  return typeof name === 'function'
+    ? wrapFn(target, name)
+    : wrapMethod(target, name, wrapper)
+}
+
+function unwrap (target, name) {
+  if (!target) return target // no target to unwrap
+
+  const unwrapper = unwrappers.get(name ? target[name] : target)
+
+  if (!unwrapper) return target // target is already unwrapped or isn't wrapped
+
+  unwrapper()
+
+  return target
+}
+
+function massWrap (targets, names, wrapper) {
+  targets = [].concat(targets)
+  names = [].concat(names)
+
+  for (const target of targets) {
+    for (const name of names) {
+      wrap(target, name, wrapper)
+    }
+  }
+}
+
+function massUnwrap (targets, names) {
+  targets = [].concat(targets)
+  names = [].concat(names)
+
+  for (const target of targets) {
+    for (const name of names) {
+      unwrap(target, name)
+    }
+  }
+}
+
+function assertMethod (target, name) {
+  if (!target) {
```

Comment:
Should types also be checked in these functions?

Reply:
Do types actually matter here? If yes we should focus only on the types that do.

Reply:
Yeah, because non-object, non-function primitives can't have properties.

Reply:
Fixed.